wasm/validation_test.cpp

```diff
@@ -0,0 +1,232 @@
+// SPDX-FileCopyrightText: 2024 David Zero <zero-one@zer0-one.net>
+//
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include "wasm/validation.h"
+
+#include "wasm/instructions.h"
+#include "wasm/types.h"
+#include "wasm/wasm.h"
+
+#include "etest/etest2.h"
+
+#include <tl/expected.hpp>
+
+#include <vector>
+
+int main() {
+    etest::Suite s{"wasm::validation"};
+
+    using namespace wasm;
+    using namespace wasm::instructions;
+    using namespace wasm::validation;
+
+    Module m{};
+    m.function_section = FunctionSection{.type_indices = {0}};
+    m.type_section = TypeSection{.types = {FunctionType{.parameters = {}, .results = {ValueType::Int32}}}};
+    m.code_section = CodeSection{.entries = {CodeEntry{}}};
+
+    CodeEntry &ce = m.code_section->entries[0];
+
+    s.add_test("Function: empty sequence", [&](etest::IActions &a) { a.expect(is_valid(m).has_value()); });
+
+    s.add_test("Function: valid trivial sequence", [&](etest::IActions &a) {
+        ce.code = {I32Const{42}, I32Const{42}, I32Add{}, I32CountLeadingZeros{}};
+
+        a.expect(is_valid(m).has_value());
+    });
+
+    s.add_test("Function: invalid trivial sequence", [&](etest::IActions &a) {
+        ce.code = {I32Const{42}, I32Add{}};
+
+        tl::expected res = is_valid(m);
+
+        a.expect(!res.has_value());
+        a.expect_eq(res.error(), ValidationError::ValueStackUnderflow);
```

Comment:
Not necessary, but you can write all of this as just `a.expect_eq(is_valid(m), tl::unexpected{ValidationError::ValueStackUnderflow});`. No need to store, check has_value, and unpack the error every time. Comparisons with tl::expected are implemented in a pretty sane fashion.